// commonutils/index.js
/**
 * Created by licp on 2016-11-28.
 */



module.exports =
{
    //common目录。
    utils      :    require('./lib/common/utils'),
    log        :    require('./lib/common/log'),
    output     :    require('./lib/common/output'),
    request   :   require('./lib/common/request'),
    statusCode     :    require('./lib/common/statusCode'),
    TimeLogger     :    require('./lib/common/TimeLogger'),
    URIRedirect     :    require('./lib/common/URIRedirect'),
    errorCodeTable     :    require('./lib/common/errorCodeTable'),
    cacheAble      :    require('./lib/common/cacheAble'),
    signToken      :    require('./lib/common/signToken'),
    errorUtils     :    require('./lib/common/errorUtils'),

    kafkaProducer      :    require('./lib/common/kafkaProducer'),

    kafkaConsumer     :    require('./lib/common/kafkaConsumer'),

    //controller目录。
    ctrl_common      :    require('./lib/controllers/common'),
    expandResources   :   require('./lib/controllers/expandResources'),
   // jwtTokenResource   :   require('./lib/controllers/jwtTokenResource'),


    //models目录。
    tables      :    require('./lib/models/basetables'),

    //proxy目录。
    proxy_common      :    require('./lib/proxy/common'),
    batchQueryUtils   :    require('./lib/proxy/batchQueryUtils'),



};
