// commonutils/lib/models/basetables.js
/**
 * Created by Administrator on 2016/9/12.
 */
"use strict";
//const knex = require('./knex').knex;
//const cache = require('../common/cacheAble');
/*const config = require('../config/config');*/

class Table {
    constructor(name,knex){
        this.name = name;
        this.knex = knex;
    };

    create(info) { return this.knex(this.name).insert(info);};
    retrieve(uuid) {return this.knex(this.name).select().where('uuid', uuid);};
    update(info) {return this.knex(this.name).update(info).where('uuid', info.uuid);};
    delete(uuid) {
        return this.knex(this.name).delete().where('uuid', uuid);
    };
    query(queryStr, offset, limit, retRaw = '*', orderBy = undefined) {
        if (orderBy) {
            return this.knex(this.name).select(retRaw).whereRaw(queryStr).offset(offset).limit(limit).orderByRaw(orderBy);
        } else {
            return this.knex(this.name).select(retRaw).whereRaw(queryStr).offset(offset).limit(limit);
        }
    };
    queryBy(queryStr,retRaw='*') { return this.knex(this.name).select(retRaw).whereRaw(queryStr);};
    getCount(queryStr) {return this.knex(this.name).whereRaw(queryStr).count('uuid as count').then(results => results[0].count);}
}

//const tableNames = ['users','departments', 'organizations'];
/*
const  tableNames = config.tables;

tableNames.map((name)=>{
    exports[name+'DB']=new Table(name);
});
*/

exports.baseTable = Table;
